cpp/src/io/csv/datetime_parser.cuh

```diff
@@ -321,43 +345,32 @@ bool extractTime(char *data, int sIdx, int eIdx, int *hour, int *minute, int *se
  *
  */
 __host__ __device__
-gdf_date32 daysSinceEpoch(int year, int month, int day)  {
-
-	static unsigned short days[12] = {0,  31,  59,  90, 120, 151, 181, 212, 243, 273, 304, 334};
-
-	// years since epoch
-	int ye = year - 1970;
-
-	// 1972 was a leap year, so how many have occurred between date and 1972? Do not include this year in count
-	int lpy = ((year - 1972 - 1) / 4);
-
-	// compute days since epoch
-	gdf_date32 days_e = (ye * 365) + lpy;
-
-	// is this a leap year?
-	if ( year % 4 == 0 && month > 2)
-		days_e++;
-
-	// months since January
-	int me = month - 01;
-
-	// days up to start of month
-	days_e += days[me];
-
-	// now just add days, but not current full days since this one is not over
-	days_e +=  day;
-
-	return days_e;
+constexpr int32_t daysSinceEpoch(int year, int month, int day) {
+
+    // More details of this formula is located in cuDF datetime_ops
+    // In brief, the calculation is split over several components:
+    //     era: a 400 year range, where the date cycle repeats exactly
+    //     yoe: year within the 400 range of an era
+    //     doy: day within the 364 range of a year
+    //     doe: exact day within the whole era
+    // The months are shifted so that Feb (possible leap day) is the last month
+    // The result subtraction of 719468 shifts start date to epoch (1970-01-01)
+    year -= (month <= 2) ? 1 : 0;
+
+    const int32_t era = (year >= 0 ? year : year - 399) / 400;
+    const int32_t yoe = year - era * 400;
+    const int32_t doy = (153 * (month + (month > 2 ? -3 : 9)) + 2)/5 + day - 1;
+    const int32_t doe = yoe * 365 + yoe/4 - yoe/100 + doy;
+
+    return era * 146097 + doe - 719468;
```

Comment:
Lots of magic numbers in this code. Do you think they could be named and replaced by constexpr variables?
Seems like a perfect use for C++1 user-defined literals. https://en.cppreference.com/w/cpp/language/user_literal

Reply:
That's a cool idea. Let me work on that, if this fix isn't urgent. There's lot of magic numbers in datetime_ops too.